rclpy/src/rclpy/_rclpy.c

```diff
@@ -2971,6 +2971,191 @@ rclpy_get_node_names_and_namespaces(PyObject * Py_UNUSED(self), PyObject * args)
 }
 
 
+static rmw_ret_t
+__convert_names_and_types(
+  rcl_names_and_types_t topic_names_and_types,
+  PyObject ** pytopic_names_and_types_ptr)
+{
+  PyObject * pytopic_names_and_types = PyList_New(topic_names_and_types.names.size);
+  *pytopic_names_and_types_ptr = pytopic_names_and_types;
+  size_t i;
+  for (i = 0; i < topic_names_and_types.names.size; ++i) {
+    PyObject * pytuple = PyTuple_New(2);
+    PyTuple_SetItem(
+      pytuple, 0,
+      PyUnicode_FromString(topic_names_and_types.names.data[i]));
+    PyObject * types_list = PyList_New(topic_names_and_types.types[i].size);
+    size_t j;
+    for (j = 0; j < topic_names_and_types.types[i].size; ++j) {
+      PyList_SetItem(
+        types_list, j,
+        PyUnicode_FromString(topic_names_and_types.types[i].data[j]));
+    }
+    PyTuple_SetItem(
+      pytuple, 1,
+      types_list);
+    PyList_SetItem(
+      pytopic_names_and_types, i,
+      pytuple);
+  }
+  rmw_ret_t ret;
+  ret = rcl_names_and_types_fini(&topic_names_and_types);
```

Comment:
It seems odd to call `rcl_names_and_types_fini` here, since this function didn't allocate the memory.  Shouldn't this be in the callers instead?

Reply:
My mistake here, will remove.